✨ Feat : 회원가입 API 추가

seunGit opened the pull request:
<!-- 
체크하려면 괄호 안에 "x"를 입력하세요. 
각 규칙은 Convention 문서에 있습니다.
PR 제목에 쓰는 prefix는 다음과 같습니다.

✨ Feat : 새로운 기능
🔨️ Refactor : 코드 리팩토링
🐎 Perf : 성능을 향상
🐛 Fix : 버그를 고칠 때
🧪 Test : 테스트 코드
🚜 Rename : 파일 이름 변경 혹은 구조를 변경
🚀 Release : 배포 / 개발 작업과 관련된 모든 것
🔥 Remove : 코드 또는 파일 제거
📚 Docs : 문서
📝 Chore : 사소한 코드 또는 언어를 변경 기타 변경사항 (빌드 스크립트 수정, 패키지 매니징 설정 등)

-->

## PR 타입
<!-- 어떤 유형의 PR인지 체크해주세요. -->

<!-- 체크하려면 괄호 안에 "x"를 입력하세요. -->
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Performance Improvement
- [ ] Bugfix
- [x] Test Code
- [ ] Code style update (formatting, local variables)
- [ ] Other... Please describe:



## 개요 
- 회원가입 기능 추가
- 구현된 코드를 테스트하는 테스트 코드 추가

## 작업 및 변경 사항 
- 회원가입 기능 구현
- Repository계층 단위 테스트 코드 추가
- Service계층 단위 테스트 코드 추가
- Controller계층 통합 테스트 코드 추가

### 테스트 결과 
- 테스트 코드 실행 특이사항 없음

close #4

seunGit committed:
[test] 회원가입기능 테스트 코드 추가.
Repository, Service 계층 단위테스트 코드 추가.
Controller 계층 은 통합테스트 진행.
회원가입 기능 테스트 완료.

Resolves: #4

seunGit committed:
[feat] 회원가입 기능 추가.
1. 회원가입 기능 추가

Resolves: #4

m1naworld commented on src/main/java/com/hanbang/e/member/service/MemberService.java:
```diff
@@ -0,0 +1,33 @@
+    @Transactional
+    public ResponseMember signup(RequestCreateMember requestCreateMember) {
+        validateDuplicateMember(requestCreateMember);
+        Member member = new Member(requestCreateMember.getEmail(), requestCreateMember.getPassword(), requestCreateMember.getAddress());
```
**엔티티 생성시** 정적팩토리를 기본으로, 파라미터가 많고 요청 객체에서 엔티티를 생성할 수 없는 경우 빌더패턴을 사용하기로 컨밴션을 정하였습니다! 컨벤션 한번 확인 해주세욥!!

seunGit replied:
컨벤션 확인하고 수정하겠습니다!

m1naworld reviewed (approved):
👏🏻👏🏻 테스트 코드까지 !! 고생하셨습니다 ! 👏🏻

hyunmuam commented on src/main/java/com/hanbang/e/member/dto/RequestCreateMember.java:
```diff
@@ -0,0 +1,34 @@
+@Builder
+public class RequestCreateMember {
+
+    @NotBlank(message = "이메일을 입력해주세요.")
+    @Email(message = "이메일 형식을 맞춰주세요.")
+    private String email;
+
+    @Pattern(regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!#%*?&])[A-Za-z\\d@$!#%*?&]{8,15}$",
+            message = "비밀번호는 8~15자리의 대소문자,숫자,특수문자로 이루어져야 합니다.")
+    private String password;
```
valid가 실패했을 때 관련된 Exception핸들러가 추가되었음 좋겠습니다. 

제가 작성한 코드는 `GlobalExceptionHandler`에 다음과 같이 `ExceptionHandler`를 추가해줬습니다.
```java
  @ExceptionHandler(MethodArgumentNotValidException.class)
  public ResponseEntity<?> validException(MethodArgumentNotValidException e) {
  String errorMsg = e.getBindingResult().getAllErrors().get(0).getDefaultMessage();
  CMResponseDto response = CMResponseDto.builder().result("fail").msg(errorMsg).data(null).build();
  return new ResponseEntity<>(response,HttpStatus.BAD_REQUEST);
  }
```

seunGit replied:
코드 확인하고 적용시켜보겠습니다!

hyunmuam commented on src/main/java/com/hanbang/e/member/dto/RequestCreateMember.java:
```diff
@@ -0,0 +1,34 @@
+
+@Getter
+@Builder
+public class RequestCreateMember {
```
저희 노션 컨벤션에 Dto 네이밍 부분 확인해주세요~

seunGit replied:
확인하겠습니다!

hyunmuam commented on src/main/java/com/hanbang/e/member/dto/ResponseMember.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@NoArgsConstructor
+public class ResponseMember {
```
마찬가지로 DTO네이밍 규칙 확인 부탁드려요.

seunGit replied:
확인하고 반영하겠습니다!

hyunmuam commented on src/main/java/com/hanbang/e/member/dto/ResponseMember.java:
```diff
@@ -0,0 +1,18 @@
+@NoArgsConstructor
+public class ResponseMember {
+
+    private String email;
+
+    private String address;
```
DTO관련 컨벤션을 보시면 저희는 DTO필드에 빈줄을 빼기로 했습니다. 확인하고 반영 부탁드립니다.

seunGit replied:
확인하고 반영하겠습니다!

hyunmuam reviewed (changes requested):
코멘트 확인 부탁드립니다.

seunGit committed:
[test] 회원가입기능 테스트 코드 수정.
Repository, Service, Controller 코드 수정.
회원가입 기능 테스트 완료.

Resolves: #4

seunGit committed:
[style] Dto 컨벤션 적용, 정적 팩토리 메서드 적용
- Dto 네이밍 규칙 적용
- Dto 네이밍 규칙 적용에 따른 import문 수정
- Dto 작성 규칙 적용(빈줄 제거)
- Dto 내 정적 팩토리 메서드 적용

seunGit committed:
[refactor] 회원가입 기능 코드 수정
- 정적 팩토리 메서드 적용에 따른 코드 수정
- Dto 네이밍 변환으로 인한 코드, import문 수정

hyunmuam commented on src/main/java/com/hanbang/e/member/dto/MemberCreateReq.java:
```diff
@@ -0,0 +1,41 @@
+            message = "비밀번호는 8~15자리의 대소문자,숫자,특수문자로 이루어져야 합니다.")
+    private String password;
+    @NotBlank(message = "주소는 필수 입력입니다.")
+    private String address;
```
와우... 이렇게보면 필드간 빈줄을 추가 해야겠다는 생각이 들긴하네요.... 이부분 한번 논의 해보시죠

seunGit replied:
이렇게 보니 또 다르네요 ㅠㅠ 확인했습니다.

hyunmuam commented on src/main/java/com/hanbang/e/member/service/MemberService.java:
```diff
@@ -0,0 +1,33 @@
+    @Transactional
+    public MemberResp signup(MemberCreateReq memberCreateReq) {
+        validateDuplicateMember(memberCreateReq);
+        Member member = Member.of(memberCreateReq.getEmail(), memberCreateReq.getPassword(), memberCreateReq.getAddress());
```
meberCreateReq에서 정의한 ToEntity를 사용하면 더 깔끔해보이긴하는데 이것도 논의해봐야 될 것 같습니다.
```java
Member member = meberCreateReq.toEntity();
```

seunGit replied:
추후 논의 후 리팩토링 여부 확인하면 좋을 것 같습니다.

hyunmuam commented on src/test/java/com/hanbang/e/member/controller/MemberControllerTest.java:
```diff
@@ -0,0 +1,89 @@
+        memberRepository.deleteAll();
+    }
+
+    @DisplayName("회원가입")
+    @Test
+    public void signupTest() throws JsonProcessingException {
+        /* given - 데이터 준비 */
+        MemberCreateReq memberCreateReq = new MemberCreateReq(
+                "hanghae4@naver.com","Hanghae11!@","부산시");
+
+        String body = om.writeValueAsString(memberCreateReq);
+
+        /* when - 테스트 실행 */
+        HttpEntity<String> request = new HttpEntity<>(body, headers);
+        ResponseEntity<String> response = rt.exchange("/api/member/signup", HttpMethod.POST, request, String.class);
+
+        /* then - 검증 */
+        DocumentContext dc = JsonPath.parse(response.getBody());
+        String result = dc.read("$.result");
+        String msg = dc.read("$.msg");
+
+        assertThat(result).isEqualTo("success");
+        assertThat(msg).isEqualTo("회원가입 완료");
+    }
```
좋은 테스트 코드 같습니다!
추후 중복된 이메일 요청, 정해진 비밀번호 형식을 지키지 않은 요청 등 처리되면 안되는 요청들이 잘 막히는지도 확인해보면 좋을 것 같네요.

hyunmuam commented on src/test/java/com/hanbang/e/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+        memberRepository.save(member2);
+    }
+
+    @AfterEach
+    public void clear() {
+        memberRepository.deleteAll();
+    }
```
현재있는 테스트는 하나이고, 테스트에서의 하나의 트랜잭션이 끝나는 시점에서 롤백되기 때문에 이 부분은 필요 없을 것 같습니다.

seunGit replied:
확인했습니다!

hyunmuam reviewed (changes requested):
고생하셨습니다. 몇가지 코멘트를 남겼는데 확인 부탁드립니다. 승기님의 코드를 보고있으니 정상적엔 요청 뿐만 아니라 처리되면 안되는 비정상적 요청에 대해서도 테스트를 해야되겠다는 생각이 드네요. 쪼금만 더 고생하시면 될 것 같습니다!

m1naworld commented on src/main/java/com/hanbang/e/member/dto/MemberCreateReq.java:
```diff
@@ -10,20 +10,27 @@
     @NotBlank(message = "주소는 필수 입력입니다.")
     private String address;
 
+    @Builder
+    public MemberCreateReq(String email, String password, String address) {
+        this.email = email;
+        this.password = password;
+        this.address = address;
+    }
+    public static MemberCreateReq from(Member member) {
+        return new MemberCreateReq(member.getEmail(), member.getPassword(), member.getAddress());
+    }
+
```
메소드 사이에는 빈 줄을 삽입을 삽입해주세요! 

저희가 따르기로한 캠퍼스 핵데이 입니다. 참고 해주세요! 🫡
https://naver.github.io/hackday-conventions-java/#blankline-between-methods

seunGit committed:
[test] 회원가입기능 테스트 코드 수정.
Repository, Service, Controller 코드 수정.
회원가입 기능 테스트 완료.

Resolves: #4

seunGit committed:
[refactor] 회원가입 기능 코드 수정
- 정적 팩토리 메서드 적용에 따른 코드 수정
- 코드 간결화 작업

Resolves: #4

hyunmuam commented on src/main/java/com/hanbang/e/member/service/MemberService.java:
```diff
@@ -0,0 +1,33 @@
+        return MemberResp.from(member);
+    }
+
+    private void validateDuplicateMember(MemberCreateReq memberCreateReq) {
+        memberRepository.findByEmail(memberCreateReq.getEmail())
+                .ifPresent(m -> {
+                    throw new IllegalArgumentException("중복된 이메일이 존재합니다.");
+                });
+    }
```
아주아주 개인적인 생각이지만 다른 메서드와 중복되는 부분이없어 아직은 따로 뺄 이유가 없어보이긴 합니다.

seunGit replied:
코드 정리해보겠습니다!

hyunmuam commented on src/test/java/com/hanbang/e/member/controller/MemberControllerTest.java:
```diff
@@ -0,0 +1,77 @@
+        headers.setContentType(MediaType.APPLICATION_JSON);
+    }
+
+    @BeforeEach
+    public void setup() {
+
+        Member member = new Member("이메일","비밀번호","주소");
+        memberRepository.save(member);
+    }
```
회원가입부분에서 쓰이고 있지 않습니다. 삭제해도 될 것 같습니다.

seunGit replied:
삭제처리 하겠습니다!

hyunmuam commented on src/test/java/com/hanbang/e/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,45 @@
+    @DisplayName("회원가입")
+    public void createMemberTest() {
+        /* given - 데이터 준비 */
+        Member member = new Member("이메일","비밀번호","주소");
```
`setup()`에서 회원을 이미 만들어주고 또 새로 만드는 이유가 없어보입니다. 다른 메소드가 없기 때문에 `BeforeEach`는 삭제해도 될 듯 합니다.

seunGit replied:
`BeforeEach` 삭제하고 반영하겠습니다.

hyunmuam commented on src/test/java/com/hanbang/e/member/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,39 @@
+        MemberResp memberResp = memberService.signup(memberCreateReq);
+
+        /* then - 검증 */
+        assertThat(memberResp.getEmail()).isEqualTo(memberCreateReq.getEmail());
```
이메일 뿐만 아니라 회원가입 시 들어간 모든 데이터를 확인해보면 좋을 것 같습니다.

seunGit replied:
`address` 데이터 추가 반영하겠습니다.

hyunmuam reviewed (changes requested):
고생하셨습니다! 코멘트 확인해주세요.

seunGit committed:
[refactor] 회원가입 기능 코드 수정
- 코드 간결화 작업

Resolves: #4

seunGit committed:
[test] 회원가입기능 테스트 코드 수정.
Repository, Controller 내 불필요 코드 제거
Service 계층 검증 목록 추가
회원가입 기능 테스트 완료.

Resolves: #4

hyunmuam reviewed (approved):
확인했습니당~

m1naworld commented on src/test/java/com/hanbang/e/member/service/MemberServiceTest.java:
```diff
@@ -2,37 +2,33 @@
 
-    @InjectMocks
+    @Autowired
     private MemberService memberService;
```
혹시 Mock 객체에서 진짜 memberService를 주입하는 코드로 변동된 이유가 있을까요??

seunGit replied:
<img width="844" alt="스크린샷 2023-01-09 오후 3 24 31" src="https://user-images.githubusercontent.com/110602191/211251355-3fab5b56-5402-41ac-a89a-2efbf153948e.png">
InjectMocks 어노테이션은 Mock 어노테이션이 붙은  객체를 주입시켜 줍니다.<br>
객체 주입을 하지 않고 적용시키려다 보니 @Autowired 를 사용한 것 같습니다. (memberRepository 누락)<br>
코멘트 확인 하고 위 그림과 같이 수정하여 반영하니 테스트가 통과되었습니다. 수정하고 더 공부하겠습니다.<br>
참고 블로그 : https://lemontia.tistory.com/951

m1naworld reviewed (commented):
고생하셨습니다 👏🏻

seunGit committed:
[test] 회원가입기능 테스트 코드 수정.
Service 계층 검증 객체 주입코드 추가
회원가입 기능 테스트 완료.

Resolves: #4
